pep-0702.rst

```diff
@@ -0,0 +1,278 @@
+==============
+
+Runtime warnings
+----------------
+
+Users might expect usage of the ``@deprecated`` decorator to issue a
+``DeprecationWarning`` at runtime. However, this would raise a number of
+thorny issues:
+
+* When the decorator is applied to a class or an overload, the warning
+  would not be raised as expected.
+* Users may want to control the ``warn`` call in more detail (e.g.,
+  changing the warning class).
+* ``typing.py`` generally aims to avoid affecting runtime behavior.
+
+Users who want to use ``@deprecated`` while also issuing a runtime warning
+can use the ``if TYPE_CHECKING:`` idiom, for example:
+
+.. code-block:: python
+
+   from typing import TYPE_CHECKING
+   import functools
+   import warnings
+
+   if TYPE_CHECKING:
+       from typing import deprecated
+   else:
+       def deprecated(msg):
+           def decorator(func):
+               @functools.wraps(func)
+               def wrapper(*args, **kwargs):
+                   warnings.warn(msg, DeprecationWarning, stacklevel=2)
+                   return func(*args, **kwargs)
+               wrapper.__deprecated__ = msg
+               return wrapper
+           return decorator
+
+Deprecation of modules and attributes
+-------------------------------------
+
+This PEP covers deprecations of classes, functions and overloads. This
+allows type checkers to detect many but not all possible deprecations.
+To evaluate whether additional functionality would be worthwhile, I
+`examined <https://gist.github.com/JelleZijlstra/ff459edc5ff0918e22b56740bb28eb8b>`__
+all current deprecations in the CPython standard library.
+
+I found:
+
+* 74 deprecations of functions, methods and classes (supported by this PEP)
+* 28 deprecations of whole modules (largely due to :pep:`594`)
+* 9 deprecations of function parameters (supported by this PEP through
+  decorating overloads)
+* 1 deprecation of a constant
+* 38 deprecations that are not easily detectable in the type system (for
+  example, for calling :func:`asyncio.get_event_loop` without an active
+  event loop)
+
+Modules could be marked as deprecated by adding a ``__deprecated__``
+module-level constant. However, the need for this is limited, and it
+is relatively easy to detect usage of deprecated modules simply by
+grepping. Therefore, this PEP omits support for whole-module deprecations.
+As a workaround, users could mark all module-level classes and functions
+with ``@deprecated``.
+
+For deprecating module-level constants, object attributes, and function
+parameters, a ``Deprecated[type, message]`` type modifier, similar to
+``Annotated`` could be added. However, this would create a new place
+in the type system where strings are just strings, not forward references,
+complicating the implementation of type checkers. In addition, my data
+show that this feature is not commonly needed.
+
+Acknowledgments
+===============
+
+A discussion in the typing-sig meetup group led to useful feedback on this
+proposal.
```

Comment:
Could you link this?

Reply:
No, it was a Zoom call.

Reply:
Okay, then perhaps that could be clarified? I read this as a discussion on the Typing-SIG mailing list. E.g.

```suggestion
A Zoom call in the typing-sig meetup group led to useful feedback on this
proposal.
```

or

```suggestion
A Zoom discussion in the typing-sig meetup group led to useful feedback on this
proposal.
```

Reply:
Pushed a change